syntax = "proto3";
package proto;

message MsgControl {
    message MsgBody {
        string title = 1;
        string content = 2;
        string url = 3;
    }

    message MsgTarget {
        repeated string mod = 1;
    }

    map<int64, MsgTarget> meta = 1;
    MsgBody body = 2;
}

message InitReq {
}

message Resp {
    enum SendStatus {
        SUCCESS = 0;
        ERROR = 1;
    }
    SendStatus status = 1;
    string detail = 2;
}

message MsgControlResp {
    map<int64, Resp> resp = 2;
}

message ModuleImportReq {
    bytes data = 1;
}

message ModuleImportResp {
    Resp resp = 1;
    repeated UrlRedirect redirect = 2;

    message UrlRedirect {
        int64 id = 1;
        string pattern = 2;
        URL_TYPE url_type = 3;
        enum URL_TYPE {
            GET = 0;
            POST = 1;
        }
    }
}

message ModuleInfo {
    string module = 1;
}

message SettingResp {
    Resp resp = 1;
    string module = 2;
    bytes data = 3;
}

message RedirectData {
    int64 id = 1;
    map<string, string> form = 2;
    bytes data = 3;
}

message RedirectResp {
    Resp resp = 1;
    string content_type = 2;
    bytes data = 3;
}

service NotifyService {
    rpc SendMessage(MsgControl) returns (MsgControlResp) {}
    rpc Init(InitReq) returns (ModuleImportResp) {}
    rpc ModuleImport(ModuleImportReq) returns (ModuleImportResp) {}
    rpc GlobalSettingRequest(ModuleInfo) returns (SettingResp) {}
    rpc UserSettingRequest(ModuleInfo) returns (SettingResp) {}
    rpc DisableModule(ModuleInfo) returns (Resp) {}
    rpc Redirect(RedirectData) returns (RedirectResp) {}
}
